manual/sphinx/user_docs/bout_options.rst

```diff
@@ -303,6 +314,34 @@ given, ``NXPE`` takes precedence and ``NYPE`` is ignored):
 
     NYPE = 1  # Set number of Y processors
 
+When choosing ``NXPE`` or ``NYPE``, they must also obey some constraints:
+
+- ``NXPE`` must be a factor of the number of grid points in the x-direction
+
+  - That is, ``(nx - 4) / NXPE`` must be an integer, assuming the usual two
+    boundary points
+
+- For simple slab grids, ``NYPE`` must be a factor of the number of grid points
+  in the y-direction
+
+  - That is, ``ny / NYPE`` must be an integer
+
+- For more general topologies, the number of points per processor ``ny / NYPE``
+  must also be a factor of the number of points in each region. For example, in
+  the usual tokamak topologies:
+
+  - in single-null there are two divertor leg and one core regions
+  - in double-null there are four divertor leg, one inner core and one outer
+    core regions
+
+Please note that here "core" means "core and adjacent SOL". See
+:ref:`sec-bout-topology` for a more detailed explanation of these regions.
+
+When BOUT++ automatically chooses ``NXPE`` and ``NYPE`` it finds all valid pairs
+which give ``total number of processors = NXPE * NYPE`` and also satisfy the
+constraints above. It then chooses the pair that makes the grid on each
+processor as close to square as possible.
```

Comment:
constraints above. It then chooses the pair that minimizes the boundaries, i.e. results ideally on a square domain on each process.

"as close to square" is not well defined (imho) - other option would be an aspect ratio close to one.

Reply:
Agreed 'as close to square' isn't great. It sounds a bit like geometry rather than number of grid points. 'aspect ratio' could too. Maybe just 'It then chooses the pair with number of points in the x- and y-directions as close to equal as possible.'?

Reply:
"Minimise the difference between NXPE and NYPE"?

Reply:
> "Minimise the difference between NXPE and NYPE"?

It's really 'minimise the difference between MXSUB and MYSUB' which isn't quite the same. I think it's equivalent to that. The actual code is
https://github.com/boutproject/BOUT-dev/blob/5461a40d963e400c2543ea6e1e1bffa018364169/src/mesh/impls/bout/boutmesh.cxx#L322
https://github.com/boutproject/BOUT-dev/blob/5461a40d963e400c2543ea6e1e1bffa018364169/src/mesh/impls/bout/boutmesh.cxx#L398-L400

Reply:
Oh yes, sorry, good point!

How about: "It then chooses the pair that makes the grid on each processor as close to square as possible (technically it chooses the pair that minimises `((nx - 4) * NPES / ny) - NXPE`, where `NPES` is total number of processors)"

Reply:
sounds good to me

Reply:
`((nx - 4) * NPES / ny) - NXPE` is minimal for maximal NXPE
You need to add `abs`

Reply:
I am even more confused - why is this linear in NPES?

The code uses a square root:

https://github.com/boutproject/BOUT-dev/blob/5461a40d963e400c2543ea6e1e1bffa018364169/src/mesh/impls/bout/boutmesh.cxx#L322

Maybe:
find NXPE as close as possible to `sqrt((nx-4) / ny * NPES)` - I like putting nx and ny together to make it clear the ratio is important. Maybe splitting  it in 2 `sqrt` makes it even more clear - part of the contribution is the number of cores, the other the ratio ...

Reply:
Oops, sorry, only just saw this. Yes, a typo there. That's what you get for trying to do two things at once!

I'll fix this directly in the v4.3.2-rc PR next week.